src/Zusi3TCP.h

```diff
@@ -215,41 +199,62 @@ namespace zusi
 			return m_id;
 		}
 
-		//! Utility function to set the value as Word
-		void setValueUint16(uint16_t value)
+		//! Utility function to set the value
+		template<typename T>
+		void setValue(T value)
 		{
-			data = new uint16_t(value);
-			data_bytes = sizeof(value);
-		}
-
-		//! Utility function to set the value as SmallInt
-		void setValueInt16(int16_t value)
-		{
-			data = new int16_t(value);
-			data_bytes = sizeof(value);
-		}
-
-		//! Utility function to set the value as Byte
-		void setValueUint8(uint8_t value)
-		{
-			data = new uint8_t(value);
-			data_bytes = sizeof(value);
-		}
-
-		//! Utility function to set the value as Single
-		void setValueFloat(float value)
-		{
-			data = new float(value);
-			data_bytes = sizeof(value);
+            static_assert(sizeof(T) <= sizeof(data), "Trying to write too big data into attribute");
+			*reinterpret_cast<T*>(data) = value;
+			data_bytes = sizeof(T);
 		}
 
+        void setValueRaw(const uint8_t *data, size_t length)
+        {
+            if (length > sizeof(this->data)) {
+                throw std::runtime_error("Trying to write too much data");
+            }
+            data_bytes = length;
+            memcpy(this->data, data, length);
+        }
+
+        bool hasValue() const
+        {
+            return data_bytes != 0;
+        }
+
+        template<typename T>
+        bool hasValueType() const
+        {
+            return sizeof(T) == data_bytes;
+        }
+
+        size_t getValueLen() const
+        {
+            return data_bytes;
+        }
+
+        template<typename T>
+        T getValue() const
+        {
+            static_assert(sizeof(T) <= sizeof(data), "Trying to read too big data into attribute");
+            return *reinterpret_cast<const T*>(data);
+        }
+
+        const uint8_t *getValueRaw(size_t *len) const
+        {
+            *len = data_bytes;
+            return data;
+        }
+
+    private:
 		//! Number of bytes in #data
 		uint32_t data_bytes;
 
 		//! Attribute data
-		void* data;
+		uint8_t data[255]; // more or less random length, this won't be enough for the timetable XML, but most likely this client
```

Comment:
I haven't tested it, and I might be mistaken about the API, but the intention was that the` MSG_WAITALL` flag in the recv call in WinsockBlockingSocket.cpp should take care of packet fragmentation.